src/UglyToad.PdfPig/DocumentLayoutAnalysis/TextEdgesExtractor.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using UglyToad.PdfPig.Content;
+using UglyToad.PdfPig.Geometry;
+
+namespace UglyToad.PdfPig.DocumentLayoutAnalysis
+{
+    /// <summary>
+    /// Text edges extractor. Text edges are where words have either there BoundingBox's left, right or mid coordinates aligned on the same vertical line.
+    /// <para>Useful to detect text columns, tables, justified text, lists, etc.</para>
+    /// </summary>
+    public class TextEdgesExtractor
+    {
+        /// <summary>
+        /// Functions used to define left, middle and right edges.
+        /// </summary>
+        private static readonly Tuple<string, Func<PdfRectangle, decimal>>[] edgesFuncs = new Tuple<string, Func<PdfRectangle, decimal>>[]
+        {
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("left",   x => Math.Round(x.Left, 0)),                // use BoundingBox's left coordinate
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("mid", x => Math.Round(x.Left + x.Width / 2, 0)),     // use BoundingBox's mid coordinate
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("right",  x => Math.Round(x.Right, 0))                // use BoundingBox's right coordinate
+        };
+
+        /// <summary>
+        /// Get the text edges.
+        /// </summary>
+        /// <param name="pageWords">The words in the page.</param>
+        /// <param name="minimumElements">The minimum number of elements to define a text edge.</param>
+        public static Dictionary<string, List<PdfLine>> GetEdges(IEnumerable<Word> pageWords, int minimumElements = 4)
```

Comment:
I think this is probably a no because `uint` and the unsigned types are kind of ugly but is it worth using a `uint` here since a negative minimumElements is probably wrong. I think the code would handle it fine so I'm not fussed.

Reply:
Because I'm an immutable nerd I'd prefer the return type here to be an `IReadOnlyDictionary<TKey, TValue>`, up to you though.

Reply:
I've added the following check:
 `if (minimumElements < 0)
{
      throw new ArgumentException("TextEdgesExtractor.GetEdges(): The minimum number of elements should be positive.", "minimumElements");
}`